<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install folium
!pip3 install wget

     |████████████████████████████████| 95 kB 1.5 MB/s eta 0:00:01


In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df = pd.read_csv(spacex_csv_file)

In [5]:
spacex_df.head()

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

Now, you can take a look at what are the coordinates for each site.


In [6]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
spacex_df.head()

Launch Site        Lat       Long  class
0  CCAFS LC-40  28.562302 -80.577356      0
1  CCAFS LC-40  28.562302 -80.577356      0
2  CCAFS LC-40  28.562302 -80.577356      0
3  CCAFS LC-40  28.562302 -80.577356      0
4  CCAFS LC-40  28.562302 -80.577356      0

In [7]:
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index = False).first()
launch_sites_df.head()

Launch Site        Lat        Long  class
0   CCAFS LC-40  28.562302  -80.577356      0
1  CCAFS SLC-40  28.563197  -80.576820      1
2    KSC LC-39A  28.573255  -80.646895      1
3   VAFB SLC-4E  34.632834 -120.610746      0

In [8]:
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610746

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [9]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location = nasa_coordinate, zoom_start = 10)

In [10]:
site_map

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [11]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius = 1000, color = '#d35400', fill = True).add_child(folium.Popup('NASA Johnson Space Center'))

# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size = (20,20),
        icon_anchor = (0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [12]:
# Site full name dictionary
site_dict = {
    'CCAFS LC-40': 'Cape Canaveral Launch Complex 40',
    'CCAFS SLC-40': 'Cape Canaveral Space Launch Complex 40',
    'KSC LC-39A': 'Kennedy Space Center Launch Complex 39',
    'VAFB SLC-4E': 'Vandenberg Space Launch Complex 4'
}

# Initial the map
launch_site_map = folium.Map(location = nasa_coordinate, zoom_start = 5)

# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for site, lat, long in launch_sites_df.values:
    circle = folium.Circle([lat, long], radius = 1000, color = 'black', fill = True)
    marker = folium.map.Marker(
        [lat, long],
        icon = DivIcon(
            icon_size = (20, 20),
            icon_anchor = (0, 0),
            html = '<div style="front-size: 12; color: :#d35400;"><b>%s</b></div>' % site
        )
    )
    
    launch_site_map.add_child(circle.add_child(folium.Popup(site_dict[site])))
    launch_site_map.add_child(marker)

In [13]:
launch_site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png" />
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [14]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [15]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [16]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
def marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'

In [17]:
spacex_df['marker_color'] = spacex_df['class'].apply(marker_color)
spacex_df.sample(n = 10)

Launch Site        Lat       Long  class marker_color
3    CCAFS LC-40  28.562302 -80.577356      0          red
41    KSC LC-39A  28.573255 -80.646895      1        green
53  CCAFS SLC-40  28.563197 -80.576820      0          red
2    CCAFS LC-40  28.562302 -80.577356      0          red
49  CCAFS SLC-40  28.563197 -80.576820      1        green
14   CCAFS LC-40  28.562302 -80.577356      0          red
15   CCAFS LC-40  28.562302 -80.577356      0          red
11   CCAFS LC-40  28.562302 -80.577356      0          red
42    KSC LC-39A  28.573255 -80.646895      1        green
6    CCAFS LC-40  28.562302 -80.577356      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [18]:
# Add marker_cluster to current site_map
marker_cluster = MarkerCluster()
launch_site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']

for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    marker = folium.Marker(
        [record['Lat'], record['Long']],
        icon = folium.Icon(color = 'white', icon_color = record['marker_color'])
    )
    marker_cluster.add_child(marker)

launch_site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png" />
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png" />
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [19]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"

mouse_position = MousePosition(
    position = 'topright',
    separator = ' Long: ',
    empty_string = 'NaN',
    lng_first = False,
    num_digits = 20,
    prefix = 'Lat:',
    lat_formatter = formatter,
    lng_formatter = formatter,
)

launch_site_map.add_child(mouse_position)
launch_site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [20]:
from math import sin, cos, sqrt, atan2, radians

In [21]:
# def calculate_distance(lat1, lon1, lat2, lon2):
def calculate_distance(coordinates1, coordinates2):
    """
    coordinates1 - tuple/list of lat1 and long1;
    coordinates2 - tuple/list of lat2 and long2;
    Example:
    lat1 = coordinates1[0]
    lon1 = coordinates1[1]
    """
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(coordinates1[0])
    lon1 = radians(coordinates1[1])
    lat2 = radians(coordinates2[0])
    lon2 = radians(coordinates2[1])

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    # wtf???????!!!
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [22]:
side_coordinates_dict = {}

for index, row in launch_sites_df.iterrows():
    side_coordinates_dict[row['Launch Site']] = [row['Lat'], row['Long']]

In [23]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [24]:
# Example
# CCAFS SLC-40 closest coastline
closest_coastline = [28.56367, -80.568]
distance_coastline = calculate_distance(side_coordinates_dict['CCAFS SLC-40'], closest_coastline)

coastline_distance_marker = folium.Marker(
    closest_coastline,
    icon = DivIcon(
        icon_size = (20, 20),
        html = '<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline)
    )
).add_child(folium.Popup('CCAFS SLC-40 Closest Coastline'))
distance_line = folium.PolyLine(locations = (side_coordinates_dict['CCAFS SLC-40'], closest_coastline), weight = 1)

launch_site_map.add_child(coastline_distance_marker)
launch_site_map.add_child(distance_line)

In [25]:
launch_site_map

*TODO:* After obtained its coordinate, create a `folium.Marker` to show the distance


In [26]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [27]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
# launch_site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png" />
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


In [28]:
def draw_distance_line(launch_site, location_coordinates, location_label):
    distance = calculate_distance(side_coordinates_dict[launch_site], location_coordinates)
    
    distance_marker = folium.Marker(
        location_coordinates,
        icon = DivIcon(
            icon_size = (20, 20),
            html = '<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance)
        )
    ).add_child(folium.Popup(location_label))
    
    distance_line = folium.PolyLine(locations = (side_coordinates_dict[launch_site], location_coordinates), weight = 1)
    
    return distance_marker, distance_line

A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png" />
</center>


In [29]:
rr_coordinates = [28.57621, -80.58626]
distance_marker, distance_line = draw_distance_line('CCAFS SLC-40', rr_coordinates, 'NASA Railroad')

launch_site_map.add_child(distance_marker)
launch_site_map.add_child(distance_line)

A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png" />
</center>


In [30]:
pkwy_coordinates = [28.56485, -80.57103]
distance_marker, distance_line = draw_distance_line('CCAFS SLC-40', pkwy_coordinates, 'Samuel C Philips Parkway')

launch_site_map.add_child(distance_marker)
launch_site_map.add_child(distance_line)

A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png" />
</center>


In [31]:
city_coordinates = [28.10076, -80.64789]
distance_marker, distance_line = draw_distance_line('CCAFS SLC-40', city_coordinates, 'Melbourne')

launch_site_map.add_child(distance_marker)
launch_site_map.add_child(distance_line)

In [32]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [33]:
launch_site = 'KSC LC-39A'
coordinates = [
    [28.58023, -80.64216],
    [28.57283, -80.65392],
    [28.57592, -80.6444],
    [28.54562, -81.33119]
]

labels = ['Coasline closest to {}'.format(launch_site), 'Railroad closest to {}'.format(launch_site), 'Road closest to {}'.format(launch_site), 'City closest to {}'.format(launch_site)]

for i in range(0, 4):
    distance_marker, distance_line = draw_distance_line(launch_site, coordinates[i], labels[i])
    launch_site_map.add_child(distance_marker)
    launch_site_map.add_child(distance_line)

In [34]:
launch_site_map

In [35]:
launch_site = 'VAFB SLC-4E'
coordinates = [
    [34.63595, -120.62504],
    [34.63566, -120.62403],
    [34.63534, -120.62324],
    [34.42766, -119.84316]
]

labels = ['Coasline closest to {}'.format(launch_site), 'Railroad closest to {}'.format(launch_site), 'Road closest to {}'.format(launch_site), 'City closest to {}'.format(launch_site)]

for i in range(0, 4):
    distance_marker, distance_line = draw_distance_line(launch_site, coordinates[i], labels[i])
    launch_site_map.add_child(distance_marker)
    launch_site_map.add_child(distance_line)

In [36]:
launch_site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


- Yes, they are (max distance is less than 2 KM);
- Yes, they are (max distance is less than 1.5 KM);
- Yes, they are (max distance is less than 1.5 KM);
- If the "certain distance" is supposed to be higher than 50 KM, then yes, they do;

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


Download a skeleton Dash app to be completed in this lab

## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01)


### Other Contributors


Joseph Santarcangelo


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description          |
| ----------------- | ------- | ---------- | --------------------------- |
| 2021-05-26        | 1.0     | Yan        | Created the initial version |


Copyright © 2021 IBM Corporation. All rights reserved.
